app/frontend/src/features/communities/CommunityPage/JoinCommunityButton.tsx

```diff
@@ -0,0 +1,71 @@
+import Button from "components/Button";
+import Snackbar from "components/Snackbar";
+import {
+  JOIN_COMMUNITY,
+  LEAVE_COMMUNITY,
+} from "features/communities/constants";
+import { Community } from "pb/communities_pb";
+import { communityKey } from "queryKeys";
+import { useMutation, useQueryClient } from "react-query";
+import { service } from "service";
+
+export default function JoinCommunityButton({
+  community,
+}: {
+  community: Community.AsObject;
+}) {
+  const queryClient = useQueryClient();
+  const join = useMutation<void, Error>(
+    () => service.communities.joinCommunity(community.communityId),
+    {
+      onSuccess() {
+        queryClient.setQueryData<Community.AsObject | undefined>(
+          communityKey(community.communityId),
+          (prevData) =>
+            prevData
+              ? {
+                  ...prevData,
+                  member: true,
+                }
+              : undefined
```

Comment:
```suggestion
          (prevData) =>
            ({
                ...prevData,
                member: true,
            })
```
The spread operator is safe to use on `undefined` on objects so don't need this extra check. Formatting on this might be a bit off so might be best to not take the patch.

(Although it errors on arrays which I didn't know before (lol js) - [there's a good Stackoverflow answer on why that's the case](https://stackoverflow.com/questions/47155141/spreading-undefined-in-array-vs-object))

Reply:
I originally tried this and got a type error, because the object might be just {member: true} which is not a Community

Reply:
Ohh the return type has to be a strictly all or nothing (undefined), fair enough 👍